File: packages/native-shell-android/src/main/kotlin/xyz/self/sdk/bridge/MessageRouter.kt
// SPDX-License-Identifier: BUSL-1.1

package xyz.self.sdk.bridge

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import java.util.UUID

class MessageRouter(
    private val sendToWebView: (js: String) -> Unit,
    private val scope: CoroutineScope = CoroutineScope(SupervisorJob() + Dispatchers.Default),
) {
    private val handlers = mutableMapOf<BridgeDomain, BridgeHandler>()
    private val json =
        Json {
            ignoreUnknownKeys = true
            encodeDefaults = true
        }

    fun register(handler: BridgeHandler) {
        handlers[handler.domain] = handler
    }

    fun onMessageReceived(
        rawJson: String,
        isTrustedSource: Boolean = true,
    ) {
        if (!isTrustedSource) {
            android.util.Log.w("BridgeRouter", "Dropped message from untrusted WebView origin")
            return
        }

        val request =
            try {
                json.decodeFromString<BridgeRequest>(rawJson)
            } catch (e: Exception) {
                android.util.Log.e("BridgeRouter", "Failed to decode request: ${e::class.simpleName}")
                return
            }
        android.util.Log.d("BridgeRouter", "Received: domain=${request.domain} method=${request.method}")

        if (request.version != BRIDGE_PROTOCOL_VERSION) {
            sendResponse(
                BridgeResponse(
                    id = UUID.randomUUID().toString(),
                    domain = request.domain,
                    requestId = request.id,
                    success = false,
                    error =
                        BridgeError(
                            code = "UNSUPPORTED_VERSION",
                            message = "Protocol version ${request.version} is not supported",
                        ),
                ),
            )
            return
        }

        val handler = handlers[request.domain]
        if (handler == null) {
            sendResponse(
                BridgeResponse(
                    id = UUID.randomUUID().toString(),
                    domain = request.domain,
                    requestId = request.id,
                    success = false,
                    error =
                        BridgeError(
                            code = "DOMAIN_NOT_FOUND",
                            message = "No handler registered for domain: ${request.domain}",
                        ),
                ),
            )
            return
        }

        scope.launch {
            try {
                val result = handler.handle(request.method, request.params)
                sendResponse(
                    BridgeResponse(
                        id = UUID.randomUUID().toString(),
                        domain = request.domain,
                        requestId = request.id,
                        success = true,
                        data = result,
                    ),
                )
            } catch (e: BridgeHandlerException) {
                sendResponse(
                    BridgeResponse(
                        id = UUID.randomUUID().toString(),
                        domain = request.domain,
                        requestId = request.id,
                        success = false,
                        error =
                            BridgeError(
                                code = e.code,
                                message = e.message,
                                details = e.details,
                            ),
                    ),
                )
            } catch (e: Exception) {
                sendResponse(
                    BridgeResponse(
                        id = UUID.randomUUID().toString(),
                        domain = request.domain,
                        requestId = request.id,
                        success = false,
                        error =
                            BridgeError(
                                code = "INTERNAL_ERROR",
                                message = e.message ?: "Unknown error",
                            ),
                    ),
                )
            }
        }
    }

    fun pushEvent(
        domain: BridgeDomain,
        event: String,
        data: JsonElement,
    ) {
        val bridgeEvent =
            BridgeEvent(
                id = UUID.randomUUID().toString(),
                domain = domain,
                event = event,
                data = data,
            )
        val eventJson = json.encodeToString(bridgeEvent)
        sendToWebView("window.SelfNativeBridge._handleEvent(${escapeForJs(eventJson)})")
    }

    private fun sendResponse(response: BridgeResponse) {
        val responseJson = json.encodeToString(response)
        android.util.Log.d("BridgeRouter", "Sending response: domain=${response.domain} success=${response.success}")
        sendToWebView("window.SelfNativeBridge._handleResponse(${escapeForJs(responseJson)})")
    }

    companion object {
        fun escapeForJs(jsonStr: String): String {
            val escaped =
                jsonStr
                    .replace("\\", "\\\\")
                    .replace("'", "\\'")
                    .replace("\n", "\\n")
                    .replace("\r", "\\r")
                    .replace("\u2028", "\\u2028")
                    .replace("\u2029", "\\u2029")
            return "'$escaped'"
        }
    }
}

File: packages/native-shell-ios/Sources/SelfNativeShell/Bridge/MessageRouter.swift
// SPDX-License-Identifier: BUSL-1.1

import Foundation

final class MessageRouter {
    private var handlers: [BridgeDomain: BridgeHandler] = [:]
    private let sendToWebView: (String) -> Void

    init(sendToWebView: @escaping (String) -> Void) {
        self.sendToWebView = sendToWebView
    }

    func register(handler: BridgeHandler) {
        handlers[handler.domain] = handler
    }

    func onMessageReceived(rawJson: String, isTrustedSource: Bool) {
        guard isTrustedSource else {
            return
        }

        guard let data = rawJson.data(using: .utf8) else {
            return
        }

        let request: BridgeRequest
        do {
            request = try JSONDecoder().decode(BridgeRequest.self, from: data)
        } catch {
            return
        }

        guard request.version == 1 else {
            let response = BridgeResponse.failure(
                request: request,
                code: "UNSUPPORTED_VERSION",
                message: "Protocol version \(request.version) is not supported"
            )
            sendResponse(response)
            return
        }

        guard let handler = handlers[request.domain] else {
            let response = BridgeResponse.failure(
                request: request,
                code: "UNKNOWN_DOMAIN",
                message: "No handler for domain: \(request.domain.rawValue)"
            )
            sendResponse(response)
            return
        }

        let params = request.params?.mapValues { $0.value }
        dispatchRequest(request, handler: handler, params: params)
    }

    private func dispatchRequest(_ request: BridgeRequest, handler: BridgeHandler, params: [String: Any]?) {
        Task {
            do {
                let result = try await handler.handle(method: request.method, params: params)
                let response = BridgeResponse.success(request: request, result: result)
                await MainActor.run { sendResponse(response) }
            } catch let error as BridgeHandlerError {
                let response = BridgeResponse.failure(
                    request: request,
                    code: error.code,
                    message: error.errorDescription ?? "Unknown error"
                )
                await MainActor.run { sendResponse(response) }
            } catch {
                let response = BridgeResponse.failure(
                    request: request,
                    code: "HANDLER_ERROR",
                    message: error.localizedDescription
                )
                await MainActor.run { sendResponse(response) }
            }
        }
    }

    func pushEvent(domain: BridgeDomain, event: String, data: Any?) {
        let eventDict: [String: Any] = [
            "type": "event",
            "version": 1,
            "domain": domain.rawValue,
            "event": event,
            "data": data as Any,
            "timestamp": Date().timeIntervalSince1970 * 1000
        ]

        guard let jsonData = try? JSONSerialization.data(withJSONObject: eventDict),
              let jsonStr = String(data: jsonData, encoding: .utf8) else {
            return
        }

        let escaped = escapeForJs(jsonStr)
        let js = "window.SelfNativeBridge._handleEvent('\(escaped)')"
        sendToWebView(js)
    }

    private func sendResponse(_ response: BridgeResponse) {
        guard let data = try? JSONEncoder().encode(response),
              let jsonStr = String(data: data, encoding: .utf8) else {
            return
        }

        let escaped = escapeForJs(jsonStr)
        let js = "window.SelfNativeBridge._handleResponse('\(escaped)')"
        sendToWebView(js)
    }

    private func escapeForJs(_ str: String) -> String {
        str.replacingOccurrences(of: "\\", with: "\\\\")
           .replacingOccurrences(of: "'", with: "\\'")
           .replacingOccurrences(of: "\n", with: "\\n")
           .replacingOccurrences(of: "\r", with: "\\r")
           .replacingOccurrences(of: "\u{2028}", with: "\\u2028")
           .replacingOccurrences(of: "\u{2029}", with: "\\u2029")
    }
}
